crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+use std::{
+    cell::RefCell,
+    collections::VecDeque,
+    hash::{BuildHasherDefault, Hasher},
+    panic::Location,
+    sync::{LazyLock, RwLock},
+};
+
+use dbsp::{
+    dynamic::{DowncastTrait, DynData},
+    operator::communication::new_exchange_operators,
+    trace::{
+        BatchReader, BatchReaderFactories, Cursor, OrdIndexedWSet as DynOrdIndexedWSet,
+        OrdIndexedWSetFactories, SpineSnapshot,
+    },
+    utils::Tup1,
+    Circuit, DynZWeight, OrdZSet, RootCircuit, Runtime, Stream, ZWeight,
+};
+use quick_cache::{
+    sync::{Cache, DefaultLifecycle, GuardResult},
+    OptionsBuilder, Weighter,
+};
+
+use crate::{SqlString, Uuid};
+
+// TODO: experiment with shard counts.
+
+/// Estimated number of cache entries used by quick_cache to provision internal resources.
+const CACHE_CAPACITY: usize = 1 << 26;
+
+/// Maximum weight of the cache, in bytes.
+///
+/// We use memory occupied by each cache entry as weight; so this roughly limits cache to 1GiB
+/// plus new entries created during the last two steps, which are pinned in the cache by setting
+/// their weight to 0.
+const WEIGHT_CAPACITY: u64 = 1 << 30;
+
+/// FIXME. We use 128-bit integers to represent interned strings.
+/// The Uuid type is the closest thing we have in SQL. Once we have compiler support for
+/// string interning we will be able to use a separate type for this.
+pub type InternedStringId = Uuid;
+
+/// String and a flag that indicates whether the string should be pinned in the cache.
+type InternedString = (SqlString, bool);
+
+/// Keys in the interned string cache are already hashes; we don't need to hash them again.
+/// This hasher just uses the first 8 bytes of the 128-bit key as the hash.
+#[derive(Default)]
+struct IdentityHasher {
+    hash: u64,
+}
+
+impl Hasher for IdentityHasher {
+    #[inline]
+    fn write(&mut self, bytes: &[u8]) {
+        debug_assert_eq!(bytes.len(), 16, "Expected 16 bytes");
+        self.hash = u64::from_le_bytes(bytes[0..8].try_into().unwrap());
+    }
+
+    #[inline]
+    fn write_u64(&mut self, i: u64) {
+        self.hash = i;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.hash
+    }
+}
+type BuildIdentityHasher = BuildHasherDefault<IdentityHasher>;
+
+/// Weighter for the interned string cache. It uses the length of the string plus the size of the
+/// key as the weight. If the string is pinned, its weight is 0, so it will not be evicted from the
+/// cache.
+#[derive(Clone)]
+struct StringWeighter;
+
+impl Weighter<InternedStringId, InternedString> for StringWeighter {
+    fn weight(&self, _key: &InternedStringId, val: &InternedString) -> u64 {
+        if val.1 {
+            0
+        } else {
+            (val.0.len() + size_of::<InternedStringId>()) as u64
+                + size_of::<InternedString>() as u64
+        }
+    }
+}
+
+thread_local! {
+    /// Current step number, used to pin recently created strings in the cache.
```

Comment:
this looks like it belongs to the RootCircuit

Reply:
True, but we don't have access to the RootCircuit from UDFs, that's why I used thread locals.